src/rowbinary/de.rs

```diff
@@ -1,94 +1,194 @@
+    pub(crate) input: &'cursor mut &'data [u8],
 }
 
-impl<'data> RowBinaryDeserializer<'_, 'data> {
-    fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
+impl<'data, Validator> RowBinaryDeserializer<'_, 'data, Validator>
+where
+    Validator: ValidateDataType,
+{
+    pub(crate) fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
         Ok(self.read_slice(size)?.to_vec())
     }
 
-    fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
+    pub(crate) fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
         ensure_size(&mut self.input, size)?;
         let slice = &self.input[..size];
         self.input.advance(size);
         Ok(slice)
     }
 
-    fn read_size(&mut self) -> Result<usize> {
+    pub(crate) fn read_size(&mut self) -> Result<usize> {
         let size = get_unsigned_leb128(&mut self.input)?;
         // TODO: what about another error?
         usize::try_from(size).map_err(|_| Error::NotEnoughData)
     }
 }
 
-#[inline]
-fn ensure_size(buffer: impl Buf, size: usize) -> Result<()> {
-    if buffer.remaining() < size {
-        Err(Error::NotEnoughData)
-    } else {
-        Ok(())
-    }
-}
-
 macro_rules! impl_num {
-    ($ty:ty, $deser_method:ident, $visitor_method:ident, $reader_method:ident) => {
+    ($ty:ty, $deser_method:ident, $visitor_method:ident, $reader_method:ident, $serde_type:expr, $type_hints:expr) => {
         #[inline]
         fn $deser_method<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
+            self.validator.validate($serde_type, $type_hints, 0)?;
             ensure_size(&mut self.input, mem::size_of::<$ty>())?;
             let value = self.input.$reader_method();
             visitor.$visitor_method(value)
         }
     };
 }
 
-impl<'data> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data> {
+impl<'data, Validator> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data, Validator>
+where
+    Validator: ValidateDataType,
+{
     type Error = Error;
 
-    impl_num!(i8, deserialize_i8, visit_i8, get_i8);
-
-    impl_num!(i16, deserialize_i16, visit_i16, get_i16_le);
-
-    impl_num!(i32, deserialize_i32, visit_i32, get_i32_le);
-
-    impl_num!(i64, deserialize_i64, visit_i64, get_i64_le);
-
-    impl_num!(i128, deserialize_i128, visit_i128, get_i128_le);
-
-    impl_num!(u8, deserialize_u8, visit_u8, get_u8);
-
-    impl_num!(u16, deserialize_u16, visit_u16, get_u16_le);
-
-    impl_num!(u32, deserialize_u32, visit_u32, get_u32_le);
-
-    impl_num!(u64, deserialize_u64, visit_u64, get_u64_le);
-
-    impl_num!(u128, deserialize_u128, visit_u128, get_u128_le);
-
-    impl_num!(f32, deserialize_f32, visit_f32, get_f32_le);
-
-    impl_num!(f64, deserialize_f64, visit_f64, get_f64_le);
+    impl_num!(
+        i8,
+        deserialize_i8,
+        visit_i8,
+        get_i8,
+        &SerdeType::I8,
+        // TODO: shall we allow deserialization from boolean?
+        &[DataTypeHint::Int8, DataTypeHint::Bool]
```

Comment:
All of these "allowed" types need double-checking, cause this is rather naive. There are cases when we can deserialize, e.g., IPv4 as a sequence of 4 bytes, or just u32 (like in our custom serde).